### DISCUSSION TITLE
It have any other `Svg` render in `Avalonia`

### DISCUSSION THREAD

**Discussion by Coloryr** [AvaloniaUI/Avalonia#13983] [Q&A] [answered]
Before that, I used `Svg Skia`'s Avalonia 'control is used to render SVG images, but suddenly `Svg.Skia` has change the open-source license to `AGPL`, and I have to switch to `Svg.Skia` to other libraries.
Are there any other implementations of SVG renderers on Avalonia?

I don't know how to make `SVG` render through `Skia`.

**Comment by thevortexcloud** [comment]
Technically, the previous versions of the library would still be MIT licensed. This is because the MIT license does not contain any provisions to revoke the license. As such you would be free to fork it and continue to use it under MIT. It's fairly common for open source projects to do things like that when the community dislikes the license changes.  However you would not be able to use any future updates from the AGPL version without also adopting the license. So you would be stuck fixing any issues yourself and you would not be able to look at the AGPL code for reference.

**Comment by kyuranger** [comment]
BTW, I never noticed it changed the license. I remember that AGPL stipulates that software using this library must be open source. But ours is commercial software and cannot be open sourced to the outside world. So, it seems I have to downgrade or use another library.

**Reply by Coloryr** [reply]
Same

**Reply by thevortexcloud** [reply]
That's not strictly true. AGPL only requires you _provide_ source code if you distribute the program to someone outside your organisation. That includes if you are using it to serve content over the network. You don't have to open source anything. If somebody who has received a copy of the software asks for the code you then have to provide it. But anybody who has not is not entitled to it.

If you are not using your software outside your company you are fine.

tl;dr Source code under GPL is for customers only

**Reply by kyuranger** [reply]
However, our software is not used within the company, but is installed in embedded devices and sold together with the devices. The relationship between the two is somewhat similar to iPhone and iOS.

**Reply by thevortexcloud** [reply]
That is a similar situation to most people (and is also why the AGPL/GPL exists). I was just pointing out you don't have to open source. It just has to be source available. You are well within your rights to indirectly paywall the source code by raising your product prices. In any case, dealing with licensing headaches like this is also why a lot of companies are very averse to using open source code. You really need a person working part time to keep track of all it and ensure license compliance.

**Reply by StefanKoell** [reply]
Wow! That's a massive showstopper for me as well. When did the license change and what was the reason. This would make it impossible to use for commercial projects, right?

I even have to reconsider using Avalonia. One reason for me to use Avalonia was the great Skia/SVG support.

**Reply by kyuranger** [reply]
BTW, this package is not an official package, but a personal package, so I don’t think it has any connection with the official Avalonia itself.

**Reply by thevortexcloud** [reply]
> This would make it impossible to use for commercial projects, right?

Nope. GPL does nothing to stop the usage of commercial products. It just imposes a lot of extra requirements (such as source code disclosure of any code that uses the GPL code).

Also as I mentioned, you could use the MIT licensed version for now. It also sounds like there would be decent demand for someone to fork it and maintain the MIT version separately. Or at least argue for the license to be changed to LGPL. Which would mean you could use the library without source disclosure as long as you don't modify the source code of the library.

**Reply by StefanKoell** [reply]
I'm not sure what package we are talking now.

Is it this?
https://www.nuget.org/packages/Svg.Skia

The owner/author is a team member of Avalonia IIRC. I'm still not sure what the motivation/need was for this "move" but license changes like this can massively hurt an ecosystem.

**Reply by thevortexcloud** [reply]
The actual source is published under their github account, not Avalonia's.

https://github.com/wieslawsoltes/Svg.Skia

**Reply by kyuranger** [reply]
Yes, this is the package we are talking about. Although the author of this package is a member of the Avalonia team, this package was not developed and released in the official capacity of Avalonia. It is a personal work.

**Reply by StefanKoell** [reply]
Well, then another package I can't use in this ecosystem. Have to restrain myself from further comments. Merry Xmas!

**Reply by kyuranger** [reply]
I am also very helpless about changing the license, but this is a fact, and I can only look for other solutions like the questioner.Merry Xmas!

**Reply by timunie** [reply]
> The owner/author is a team member of Avalonia IIRC.

No he is not a team member but a MVP and a valuable contributor.

**Reply by StefanKoell** [reply]
Good to know. Thanks!

**Comment by StefanKoell** [comment]
> Nope. GPL does nothing to stop the usage of commercial products. It just imposes a lot of extra requirements (such as source code disclosure of any code that uses the GPL code).

I'm talking practically here. In theory you are right, of course but at least for me it's a showstopper. Going back to the MIT licensed version is a band-aid at best. This is not sustainable/maintainable.

I guess I have to look for alternatives. That's really bad news.

**Comment by thevortexcloud** [comment]
In any case, it looks like Skia itself actually has SVG support.  Skiasharp can parse SVG path data. 

https://learn.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/curves/path-data

Avalonia (and also WPF) can also do this too. EG https://docs.avaloniaui.net/docs/reference/controls/detailed-reference/path-icon

But this does mean you need to convert the SVG into a path first.

**Reply by thevortexcloud** [reply]
Also the library that Svg.Skia uses has not changed their license. It can also be used to render an SVG to a bitmap (although it relies on [System.Drawing](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/system-drawing-common-windows-only) which has so so support on non Windows platforms).

https://github.com/svg-net/SVG

**Reply by thevortexcloud** [reply]
I also found this fork of the library which uses Skia directly too. It seems like it still gets somewhat maintained. 

https://github.com/gentledepp/SVG

It does however seem to be missing proper Linux support.

**Reply by StefanKoell** [reply]
Found this: https://github.com/mono/SkiaSharp.Extended/

Maybe a viable alternative - especially when it comes to cross platform support

**Reply by kyuranger** [reply]
Most of these are no longer updated, meaning they cannot enjoy the features and performance improvements of the latest version of SkiaSharp.

**Comment by timunie** [comment]
I honestly am not sure we want to discuss the decision of @wieslawsoltes here. I'm sure has a a good reason for it. But as pointed out earlier, it is not really tied or related to Avalonia.

**Reply by StefanKoell** [reply]
Well, of course, it's up to @wieslawsoltes  as it is his baby but I think I can speak for everyone when I say it's really disrupting when you invest a massive amount of time adopting a library like this into your app and learn that the license will not be usable in the future anymore. I wouldn't even look at the library when it's (x)GPL whatever but this library was MIT licensed in the beginning and now, I have to invest a lot of time (and probably money) to get this out of my codebase again.

Yes, I know, I can use the MIT version for now or fork the MIT release myself and yes I know these are the realities of OSS but it doesn't suck less because of that. I don't even mind to buy a commercial license for SVG rendering (if you have any recommendations). I'm just complaining that this is really disruptive and leaves a bad taste in my mouth.

**Reply by timunie** [reply]
Still wrong repro to discuss this. I'll limit this ticket for now.

**Reply by workgroupengineering** [reply]
> Well, of course, it's up to @wieslawsoltes as it is his baby but I think I can speak for everyone when I say it's really disrupting when you invest a massive amount of time adopting a library like this into your app and learn that the license will not be usable in the future anymore. I wouldn't even look at the library when it's (x)GPL whatever but this library was MIT licensed in the beginning and now, I have to invest a lot of time (and probably money) to get this out of my codebase again.
> 
> Yes, I know, I can use the MIT version for now or fork the MIT release myself and yes I know these are the realities of OSS but it doesn't suck less because of that. I don't even mind to buy a commercial license for SVG rendering (if you have any recommendations). I'm just complaining that this is really disruptive and leaves a bad taste in my mouth.


You're right, but you can't live on OS alone. Many large companies exploit the time dedicated by open source developers by making large profits without contributing to the project or through donations.

**Reply by StefanKoell** [reply]
@workgroupengineering maybe you can enlighten me what benefits a repo gains which solves the issues you mentioned when switching from MIT to AGPL. Maybe I'm missing something.

But as @timunie mentioned, we should shift the discussion to search for alternatives. It sucks that the license has changed and we have to live with that. Let's find a solid alternative (OSS or commercial) which can easily be used with Avalonia so that we all can move on.

**Reply by workgroupengineering** [reply]
Because companies cannot use this library unless they make their software open so contribute community efforts.

**Reply by StefanKoell** [reply]
@workgroupengineering well, that is the exact argument against it. Companies (like mine) simply cannot open source the software because it is my asset I have to earn money and make a living. Bigger companies can't simply open source their internal tools. The result is that the library will be abandoned and an alternative is used instead. A classic lose-lose situation.

**Reply by StefanKoell** [reply]
An additional note: SVG is a "defacto standard" nowadays and many UI frameworks support this out of the box/natively. I was always a big fan of Avalonia because of the excellent cross-platform support in this area. Also thanks to the work of @wieslawsoltes 

Tbh, SVG support should be included in Avalonia and I was always OK with using an "additional" library to gain that support. Now the situation has drastically changed. With this landscape I'm not sure I would go the Avalonia route again. I already have invested a lot of time and effort in an app based on Avalonia which I need to sell in order to make a living. I even bought an Avalonia support plan to financially support the project. Now, without good SVG support/integration, which I really need for my app, I'm not sure what to do.

**Comment by wieslawsoltes** [comment, marked as answer]
> Before that, I used `Svg Skia`'s Avalonia 'control is used to render SVG images, but suddenly `Svg.Skia` has change the open-source license to `AGPL`, and I have to switch to `Svg.Skia` to other libraries. Are there any other implementations of SVG renderers on Avalonia?
> 
> I don't know how to make `SVG` render through `Skia`.

If you need Svg in Avalonia you can use Svg to Xaml tool specifically built for Avalonia https://github.com/wieslawsoltes/SvgToXaml

**Reply by StefanKoell** [reply]
@wieslawsoltes so I guess the change from MIT to AGPL will not be reconsidered, correct?

**Reply by wieslawsoltes** [reply]
I might do dual-licensing for open source projects

**Reply by Coloryr** [reply]
ok thanks

**Reply by StefanKoell** [reply]
What does that mean? I think open source projects will not have an issue with AGPL per se.

**Reply by bazsisz** [reply]
@wieslawsoltes SvgToXaml needs some expplanation how to use correctly. I spent 1.5 hours trying to figure it out withou success. Would be nice if you could write some examples in README. Thanks

**Comment by thevortexcloud** [comment]
FYI for anybody following this, the license change has been reverted.

https://github.com/wieslawsoltes/Svg.Skia/commit/735eab9e7459edbf2887512818baf7c923987b9d

So for now at least, you can all continue using the library the same way you have always done.

**Comment by PanXiangquan** [comment]
There are many SVG primitives in an SVG diagram. I need to set different colors for SVG primitives. What should I do?
![Uploading PixPin_2024-05-27_11-51-48.png…]()

**Reply by kyuranger** [reply]
I have a strange way that can to solve this: 
https://github.com/AvaloniaUI/Avalonia/discussions/15759#discussioncomment-9502005

You can bind the color you like this way. Now I use this in my projects.
